### PR TITLE
upstream-analysis-util (L2/4)

### LINKED ISSUE #18: Upstream analysis module - add qc metrics function
## What are the scientific goals of the analysis?

This is to add the `function-calculate-qc-metrics.R` in the upstream-analysis module.

### PR DESCRIPTION
**Pull request by AntoniaChroni**
# Purpose/implementation Section

## Type of change

- [x] New feature (non-breaking change which adds functionality)

## What scientific question/module of analysis is your analysis addressing?

To add the `function-run-QC.R` in the upstream-analysis module.

## What GitHub issue does your pull request address?

Closes #18.

## Directions for reviewers 
Tell potential reviewers what kind of feedback you are soliciting.

### Which areas should receive a particularly close look?

Please review the code for logic and clarity.

### CODE REVIEW

File: analyses/upstream-analysis/util/function-run-QC.R [hunk 1]
```diff
@@ -0,0 +1,66 @@
...
+
+###########################################################################
+#' Function to run QC with default parameters
+#' @param sce
+#'
+#' @return
+#' @export
+#'
+#' @examples
+run_QC_default <- function(sce) {
```
Comment (sharonfreshour) [inline analyses/upstream-analysis/util/function-run-QC.R:38]: Perhaps I am misunderstanding how this function will be called, but based on the code it seems like parameters for use_only_step1 and seurat_obj should be passed to the function?

Comment (sharonfreshour) [inline analyses/upstream-analysis/util/function-run-QC.R:38]: It is a bit unclear why the function is called on sce as the parameter name, but it appears sce is not used within the function? Rather it seems like everything is done on the object called seurat_obj?

File: analyses/upstream-analysis/util/function-run-QC.R [hunk 2]
```diff
@@ -0,0 +1,66 @@
...
+      cat("Step 1 data filter", sample_name[i], "\n")
+      filtered_sce <- subset(seurat_obj, subset = nFeature_RNA >= min_genes &
+                                                  nCount_RNA >= min_count &
+                                                  percent.mito <= mtDNA_pct_default)                                   
+      # Step 2
+      cat("Step 2 data filter", sample_name[i], "\n")
+      nFeature_RNA_rm_outlier <- Find_Outlier_Thershold(seurat_obj@meta.data$nFeature_RNA)
+      nCount_RNA_rm_outlier <- Find_Outlier_Thershold(seurat_obj@meta.data$nCount_RNA)
+      mtDNA_pct_rm_outlier <- Find_Outlier_Thershold(seurat_obj@meta.data$percent.mito)
+      filtered_sce <- subset(seurat_obj,
```
Comment (sharonfreshour) [inline analyses/upstream-analysis/util/function-run-QC.R:59]: I think that the way the core has used this MAD filtering in the past is by applying it on the object that has already been filtered once by the step 1 filtering. However, it looks like this code is using the unfiltered seurat_obj for the MAD filtering and overwriting the filtered_sce object from the step above. I think we should ask @CodyRamirez for clarification on how we should use the MAD filtering and then double check this code is filtering it in that way.

**Review by sharonfreshour — COMMENTED** [review]
Waiting for clarification from Cody before approving this PR

File: analyses/upstream-analysis/util/function-run-QC.R (hunk 1 above)
Comment (AntoniaChroni) [reply analyses/upstream-analysis/util/function-run-QC.R:38]: We will use the `miQC` R package that works on `sce` objects for filtering. Therefore, the object was named accordingly. We convert back to a `seurat_obj` after that step in the QC script.

Comment (sharonfreshour) [reply analyses/upstream-analysis/util/function-run-QC.R:38]: Sorry, I should have been more clear in my comment. I see that the structure of the object used for the miQC function is an sce object. And the object structure for this function is a seurat object. I was referring more so to the naming and what the purpose of the sce parameter is. sce (the named input) is not used in this function, so it seems what the input the function is called with isn't relevant to what actual object the function is run with. It seems the function is being run with whatever object named seurat_obj exists outside the function, if that makes sense.

File: analyses/upstream-analysis/util/function-run-QC.R (hunk 2 above)
Comment (AntoniaChroni) [reply analyses/upstream-analysis/util/function-run-QC.R]: There was a misunderstanding as for the way filtering is performed. This is correct. The issue was the object being called. This has been now addressed here b5c9ec6e722c029bf1b683e698ed59c94e3d63f2.

**Review by sharonfreshour — APPROVED** [review]
function-run-QC.R looks great, code is clear and logical!

**Pull request merged**